wdl-lint/src/rules/runtime_section_keys.rs

```diff
@@ -0,0 +1,355 @@
+    /// Link: https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#runtime-section
+    static ref V1_0_KEYS: HashSet<Key> = {
+        let mut keys = HashSet::new();
+        keys.insert(Key::Recommended("docker"));
+        keys.insert(Key::Recommended("memory"));
+        keys
+    };
+
+    /// The mapping between `runtime` keys and their kind for WDL v1.1.
+    ///
+    /// Link: https://github.com/openwdl/wdl/blob/wdl-1.1/SPEC.md#runtime-section
+    static ref V1_1_KEYS: HashSet<Key> = {
+        let mut keys = HashSet::new();
+        keys.insert(Key::ReservedMandatory("container"));
+        keys.insert(Key::Deprecated(DeprecatedKey{ name: "docker", replacement: "container" }));
+        keys.insert(Key::ReservedMandatory("cpu"));
+        keys.insert(Key::ReservedMandatory("memory"));
+        keys.insert(Key::ReservedMandatory("gpu"));
+        keys.insert(Key::ReservedMandatory("disks"));
+        keys.insert(Key::ReservedMandatory("maxRetries"));
+        keys.insert(Key::ReservedMandatory("returnCodes"));
+        keys.insert(Key::Hint("maxCpu"));
+        keys.insert(Key::Hint("maxMemory"));
+        keys.insert(Key::Hint("shortTask"));
+        keys.insert(Key::Hint("localizationOptional"));
+        keys.insert(Key::Hint("inputs"));
+        keys.insert(Key::Hint("outputs"));
+        keys
+    };
+}
+
+/// Creates a "deprecated runtime key" diagnostic.
+fn deprecated_runtime_key(key: &DeprecatedKey, span: Span) -> Diagnostic {
+    Diagnostic::warning(format!(
+        "the `{}` runtime key has been deprecated in favor of `{}`",
+        key.name, key.replacement
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix(format!(
+        "change the name of the `{}` key to `{}`",
+        key.name, key.replacement
+    ))
+}
+
+/// Creates an "non-reserved runtime key" diagnostic.
+fn non_reserved_runtime_key(key: &str, span: Span, specification: &str) -> Diagnostic {
+    Diagnostic::warning(format!(
+        "the `{key}` runtime key is not reserved in {specification}; therefore, its inclusion in \
+         the `runtime` section is deprecated"
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix(
+        "if a reserved key name was intended, correct the spelling; otherwise, remove the key",
+    )
+}
+
+/// Creates a "missing recommended runtime key" diagnostic.
+fn missing_recommended_key(key: &str, span: Span, specification: &str) -> Diagnostic {
+    Diagnostic::note(format!(
+        "the `{key}` runtime key is recommended by {specification}"
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix(format!(
+        "include an entry for the `{key}` key in the `runtime` section"
+    ))
+}
+
+/// Inspects a runtime key for issues within v1.1 WDL documents.
+fn inspect_v1_1_key(state: &mut Diagnostics, key: &str, span: Span, specification: &str) {
+    let entries = V1_1_KEYS
+        .iter()
+        .filter(|entry| entry.name() == key)
+        .collect::<Vec<_>>();
```

Comment:
As `inspect_v1_1_key` is called once per runtime item, is there a way we could avoid the `collect` here?

Such as:

```rust
let mut entries = V1_1_KEYS.iter().filter(|e| e.name() == key);
let entry = entries.next();
assert!(entries.next().is_none(), "more than one key matched; duplicate entry in static key map");
match entry {
...
}
```

Also, it seems like the key sets could instead be a `HashMap<&'static str, Key>` where `Key` would be:

```rust
enum Key {
    /// A key that is deprecated in favor of another key.
    /// The value is the replacement key.
    Deprecated(&'static str),
    /// A runtime key that is recommended to be included.
    Recommended,
    /// A runtime key that has a reserved meaning in the specification but which
    /// execution engines are _not_ required to support. These are also called
    /// "hints" in WDL parlance.
    Hint,
    /// A runtime key that has a reserved meaning in the specification and which
    /// execution engines are required to support (but don't necessarily have to
    /// be present in WDL documents).
    ReservedMandatory,
}
```

That way we don't need to iterate the collection at all.

Reply:
So I originally had it the way you describe above and ended up flipping back and forth. There were some pros I liked in terms of the clarity of the code with the `HashSet` option (from the top of my head, duplicates keys would be able to be detected rather than re-inserting over an existing key + pulling out the `DeprecatedKey` made it possible to statically verify only deprecated keys could be passed to `deprecated_runtime_key` for future implementers).

All that said, I made those decisions because I felt the performance hit was small compared to a small bump up in clarity of the code but, as you and I discussed, we can take the philosophy of eeking out every bit of performance here.

Reply:
If detecting duplicates on init is desired, I think we could structure initialization to look like:

```rust
V1_1_KEYS.get_or_init(|| {
    const KEYS: &[(&str, Key)] = [("container", Key::ReservedMandatory), ...];

    let mut keys = HashMap::with_capacity(KEYS.len());
    for (k, v) in KEYS {
        let prev = keys.insert(k, v);
        assert!(prev.is_none(), "duplicate key `{k}`");
    }
    
    keys
});
```